zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,165 @@
+import org.apache.zookeeper.server.auth.X509AuthenticationProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A ServerAuthenticationProvider implementation that does both authentication and authorization for protecting znodes from unauthorized access.
+ * Znodes are grouped into domains according to their ownership, and clients are granted access permission to domains.
+ * Authentication mechanism is same as in X509AuthenticationProvider.
+ * Authorization is done by checking with clients' URI (uniform resource identifier) in ACL metadata for matched domains.
+ */
+public class X509ZNodeGroupAclProvider extends ServerAuthenticationProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(X509ZNodeGroupAclProvider.class);
+  private final String logStrPrefix = this.getClass().getName() + ":: ";
+  private final X509TrustManager trustManager;
+  private final X509KeyManager keyManager;
+
+  public X509ZNodeGroupAclProvider()
+      throws X509Exception.KeyManagerException, X509Exception.TrustManagerException {
+    // Reuse logic in X509AuthenticationProvider
+    X509AuthenticationProvider authProvider = new X509AuthenticationProvider();
+    this.keyManager = authProvider.getKeyManager();
+    this.trustManager = authProvider.getTrustManager();
+  }
+
+  public X509ZNodeGroupAclProvider(X509TrustManager trustManager, X509KeyManager keyManager) {
+    this.trustManager = trustManager;
+    this.keyManager = keyManager;
+  }
+
+  @Override
+  public KeeperException.Code handleAuthentication(ServerObjs serverObjs, byte[] authData) {
+    // Get x509 certificate
+    ServerCnxn cnxn = serverObjs.getCnxn();
+    X509Certificate[] certChain = (X509Certificate[]) cnxn.getClientCertificateChain();
+    if (certChain == null || certChain.length == 0) {
+      LOG.error(logStrPrefix + "No x509 certificate is found.");
+      return KeeperException.Code.AUTHFAILED;
+    }
+    X509Certificate clientCert = certChain[0];
+
+    if (trustManager == null) {
+      LOG.error(logStrPrefix + "No trust manager available to authenticate session 0x{}",
+          Long.toHexString(cnxn.getSessionId()));
+      return KeeperException.Code.AUTHFAILED;
+    }
+
+    try {
+      // Authenticate client certificate
+      trustManager.checkClientTrusted(certChain, clientCert.getPublicKey().getAlgorithm());
+    } catch (CertificateException ce) {
+      LOG.error(logStrPrefix + "Failed to trust certificate for session 0x{}",
+          Long.toHexString(cnxn.getSessionId()), ce);
+      return KeeperException.Code.AUTHFAILED;
+    }
+
+    // Extract URI from certificate
+    String clientId;
+    try (ClientX509Util x509Util = new ClientX509Util()) {
+      clientId = x509Util.getClientId(clientCert);
+    } catch (Exception e) {
+      // Failed to extract URI from certificate
+      LOG.error(logStrPrefix + "Failed to extract URI from certificate for session 0x{}",
+          Long.toHexString(cnxn.getSessionId()), e);
```

Comment:
Do we need a more descriptive error log? Or is it handled by `getClientId()`?

Reply:
It is handled by getClientId()